pkg/permissionclaim/permissionclaim_labeler.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package permissionclaim
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/clusters"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1/permissionclaims"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+)
+
+// Labeler calculates labels to apply to all instances of a cluster-group-resource based on permission claims.
+type Labeler struct {
+	listAPIBindingsAcceptingClaimedGroupResource func(clusterName logicalcluster.Name, groupResource schema.GroupResource) ([]*apisv1alpha1.APIBinding, error)
+	getAPIExport                                 func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error)
+}
+
+// NewLabeler returns a new Labeler.
+func NewLabeler(
+	apiBindingInformer apisinformers.APIBindingInformer,
+	apiExportInformer apisinformers.APIExportInformer,
+) *Labeler {
+	return &Labeler{
+		listAPIBindingsAcceptingClaimedGroupResource: func(clusterName logicalcluster.Name, groupResource schema.GroupResource) ([]*apisv1alpha1.APIBinding, error) {
+			indexKey := fmt.Sprintf("%s|%s", clusterName, groupResource)
+			return indexers.ByIndex[*apisv1alpha1.APIBinding](apiBindingInformer.Informer().GetIndexer(), indexers.APIBindingByClusterAndAcceptedClaimedGroupResources, indexKey)
+		},
+
+		getAPIExport: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error) {
+			key := clusters.ToClusterAwareKey(clusterName, name)
+			return apiExportInformer.Lister().Get(key)
+		},
+	}
+}
+
+// LabelsFor returns all the applicable labels for the cluster-group-resource relating to permission claims. This is
+// the intersection of (1) all APIBindings in the cluster that have accepted claims for the group-resource with (2)
+// associated APIExports that are claiming group-resource.
+func (l *Labeler) LabelsFor(_ context.Context, cluster logicalcluster.Name, groupResource schema.GroupResource) (map[string]string, error) {
```

Comment:
why an unused context?

Reply:
or in other words: if  this core helper is worth a context, it is very likely a performance bottleneck.

Reply:
i can remove it. was adding it just in case we needed it in the future

Reply:
Kept it for logging